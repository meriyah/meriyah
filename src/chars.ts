/**
 * A list of character constants with much more human-readable names.
 */
export const enum Chars {

    Null = 0x00,

    Backspace      = 0x08,
    Tab            = 0x09,
    LineFeed       = 0x0A,
    VerticalTab    = 0x0B,
    FormFeed       = 0x0C,
    CarriageReturn = 0x0D,

    Space       = 0x20,
    Exclamation = 0x21,
    DoubleQuote = 0x22,
    Hash        = 0x23,
    Dollar      = 0x24,
    Percent     = 0x25,
    Ampersand   = 0x26,
    SingleQuote = 0x27,
    LeftParen   = 0x28,
    RightParen  = 0x29,
    Asterisk    = 0x2A,
    Plus        = 0x2B,
    Comma       = 0x2C,
    Hyphen      = 0x2D,
    Period      = 0x2E,
    Slash       = 0x2F,

    Zero         = 0x30,
    One          = 0x31,
    Two          = 0x32,
    Three        = 0x33,
    Four         = 0x34,
    Five         = 0x35,
    Six          = 0x36,
    Seven        = 0x37,
    Eight        = 0x38,
    Nine         = 0x39,
    Colon        = 0x3A,
    Semicolon    = 0x3B,
    LessThan     = 0x3C,
    EqualSign    = 0x3D,
    GreaterThan  = 0x3E,
    QuestionMark = 0x3F,

    UpperA = 0x41,
    UpperB = 0x42,
    UpperC = 0x43,
    UpperD = 0x44,
    UpperE = 0x45,
    UpperF = 0x46,
    UpperG = 0x47,
    UpperH = 0x48,
    UpperI = 0x49,
    UpperJ = 0x4A,
    UpperK = 0x4B,
    UpperL = 0x4C,
    UpperM = 0x4D,
    UpperN = 0x4E,
    UpperO = 0x4F,

    UpperP       = 0x50,
    UpperQ       = 0x51,
    UpperR       = 0x52,
    UpperS       = 0x53,
    UpperT       = 0x54,
    UpperU       = 0x55,
    UpperV       = 0x56,
    UpperW       = 0x57,
    UpperX       = 0x58,
    UpperY       = 0x59,
    UpperZ       = 0x5A,
    LeftBracket  = 0x5B,
    Backslash    = 0x5C,
    RightBracket = 0x5D,
    Caret        = 0x5E,
    Underscore   = 0x5F,

    Backtick = 0x60,
    LowerA   = 0x61,
    LowerB   = 0x62,
    LowerC   = 0x63,
    LowerD   = 0x64,
    LowerE   = 0x65,
    LowerF   = 0x66,
    LowerG   = 0x67,
    LowerH   = 0x68,
    LowerI   = 0x69,
    LowerJ   = 0x6A,
    LowerK   = 0x6B,
    LowerL   = 0x6C,
    LowerM   = 0x6D,
    LowerN   = 0x6E,
    LowerO   = 0x6F,

    LowerP      = 0x70,
    LowerQ      = 0x71,
    LowerR      = 0x72,
    LowerS      = 0x73,
    LowerT      = 0x74,
    LowerU      = 0x75,
    LowerV      = 0x76,
    LowerW      = 0x77,
    LowerX      = 0x78,
    LowerY      = 0x79,
    LowerZ      = 0x7A,
    LeftBrace   = 0x7B,
    VerticalBar = 0x7C,
    RightBrace  = 0x7D,
    Tilde       = 0x7E,
    Delete      = 0x7F,

    NextLine         = 0x85,
    NonBreakingSpace = 0xA0,

    Ogham = 0x1680,

    EnQuad             = 0x2000,
    EmQuad             = 0x2001,
    EnSpace            = 0x2002,
    EmSpace            = 0x2003,
    ThreePerEmSpace    = 0x2004,
    FourPerEmSpace     = 0x2005,
    SixPerEmSpace      = 0x2006,
    FigureSpace        = 0x2007,
    PunctuationSpace   = 0x2008,
    ThinSpace          = 0x2009,
    HairSpace          = 0x200A,
    ZeroWidthSpace     = 0x200B,
    ZeroWidthJoiner    = 0x200C,
    ZeroWidthNonJoiner = 0x200D,

    LineSeparator      = 0x2028,
    ParagraphSeparator = 0x2029,

    NarrowNoBreakSpace = 0x202F,
    MathematicalSpace  = 0x205F,
    IdeographicSpace   = 0x3000,

    ZeroWidthNoBreakSpace = 0xFEFF,

    ByteOrderMark = 0xFFEF,

    NonBMPMax = 0x10FFFF,

    LeadSurrogateMin = 0xD800,
    LeadSurrogateMax = 0xDBFF,
    TrailSurrogateMin = 0xDC00,
    TrailSurrogateMax = 0xDFFF,
    UTF16Max = 0xFFFF,
    NonBMPMin = 0x10000,
}
