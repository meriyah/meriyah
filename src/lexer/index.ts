export { scanSingleToken, nextToken } from './scan';
export { skipHashBang } from './comments';
export { advanceChar, consumeLineFeed, scanNewLine, LexerState, convertTokenType } from './common';
