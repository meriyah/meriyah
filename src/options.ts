import type * as ESTree from './estree';
import { pushComment, pushToken } from './parser/parser';
import { type Token } from './token';

/**
 * Function calls when semicolon inserted.
 */
type OnInsertedSemicolon = (pos: number) => any;

/**
 * Token process function.
 */
export type OnToken = (token: string, start: number, end: number, loc: ESTree.SourceLocation) => any;

/**
 * Comment process function.
 */
export type OnComment = (
  type: ESTree.CommentType,
  value: string,
  start: number,
  end: number,
  loc: ESTree.SourceLocation,
) => any;

/**
 * The parser options.
 */
export interface Options {
  // Allow module code
  module?: boolean;
  // Enable stage 3 support (ESNext)
  next?: boolean;
  // Enable start and end offsets to each node
  ranges?: boolean;
  // Enable web compatibility
  webcompat?: boolean;
  // Enable line/column location information to each node
  loc?: boolean;
  // Attach raw property to each literal and identifier node
  raw?: boolean;
  // Allow return in the global scope
  globalReturn?: boolean;
  // Enable implied strict mode
  impliedStrict?: boolean;
  // Enable non-standard parenthesized expression node
  preserveParens?: boolean;
  // Enable lexical binding and scope tracking
  lexical?: boolean;
  // Adds a source attribute in every nodeâ€™s loc object when the locations option is `true`
  source?: string;
  // Enable React JSX parsing
  jsx?: boolean;
  // Allows comment extraction. Accepts either a callback function or an array
  onComment?: ESTree.Comment[] | OnComment;
  // Allows detection of automatic semicolon insertion. Accepts a callback function that will be passed the character offset where the semicolon was inserted
  onInsertedSemicolon?: OnInsertedSemicolon;
  // Allows token extraction. Accepts either a callback function or an array
  onToken?: Token[] | OnToken;
}

export type NormalizedOptions = Omit<Options, 'onComment' | 'onToken'> & {
  onComment?: OnComment;
  onToken?: OnToken;
};

export function normalizeOptions(rawOptions: Options): NormalizedOptions {
  const options = { ...rawOptions } as NormalizedOptions;

  // Accepts either a callback function to be invoked or an array to collect comments (as the node is constructed)
  if (options.onComment) {
    options.onComment = Array.isArray(options.onComment) ? pushComment(options.onComment, options) : options.onComment;
  }

  // Accepts either a callback function to be invoked or an array to collect tokens
  if (options.onToken) {
    options.onToken = Array.isArray(options.onToken) ? pushToken(options.onToken, options) : options.onToken;
  }

  return options;
}
