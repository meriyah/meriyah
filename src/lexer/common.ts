import { Chars } from '../chars';
import { Flags } from '../common';
import { type Parser } from '../parser/parser';
import { Token } from '../token';

export const enum LexerState {
  None = 0,
  NewLine = 1 << 0,
  LastIsCR = 1 << 2,
}

export const enum NumberKind {
  ImplicitOctal = 1 << 0,
  Binary = 1 << 1,
  Octal = 1 << 2,
  Hex = 1 << 3,
  Decimal = 1 << 4,
  NonOctalDecimal = 1 << 5,
  Float = 1 << 6,
  ValidBigIntKind = 1 << 7,
  DecimalNumberKind = Decimal | NonOctalDecimal,
}

/**
 * Advances this lexer's current index.
 *
 * @param parser The parser instance
 */
export function advanceChar(parser: Parser): number {
  parser.column++;
  return (parser.currentChar = parser.source.charCodeAt(++parser.index));
}

/**
 * Consumes possible surrogate pair, return the merged code point or 0
 *
 * @param parser The parser instance
 */
export function consumePossibleSurrogatePair(parser: Parser): number {
  // See: https://tc39.github.io/ecma262/#sec-ecmascript-language-types-string-type
  const hi = parser.currentChar;
  // Not surrogate
  if ((hi & 0xfc00) !== Chars.LeadSurrogateMin) return Chars.Null;
  const lo = parser.source.charCodeAt(parser.index + 1);
  // Unpaired surrogate
  if ((lo & 0xfc00) !== Chars.TrailSurrogateMin) return Chars.Null;
  // Merged code point
  return Chars.NonBMPMin + ((hi & 0x3ff) << 10) + (lo & 0x3ff);
}

/**
 * Use to consume a line feed instead of `scanNewLine`.
 */
export function consumeLineFeed(parser: Parser, state: LexerState): void {
  parser.currentChar = parser.source.charCodeAt(++parser.index);
  parser.flags |= Flags.NewLine;
  if ((state & LexerState.LastIsCR) === 0) {
    parser.column = 0;
    parser.line++;
  }
}

export function scanNewLine(parser: Parser): void {
  parser.flags |= Flags.NewLine;
  parser.currentChar = parser.source.charCodeAt(++parser.index);
  parser.column = 0;
  parser.line++;
}

// ECMA-262 11.2 White Space
// gC=Zs, U+0009, U+000B, U+000C, U+FEFF
export function isExoticECMAScriptWhitespace(ch: number): boolean {
  return (
    ch === Chars.NonBreakingSpace ||
    ch === Chars.ZeroWidthNoBreakSpace ||
    ch === Chars.NextLine ||
    ch === Chars.Ogham ||
    (ch >= Chars.EnQuad && ch <= Chars.ZeroWidthSpace) ||
    ch === Chars.NarrowNoBreakSpace ||
    ch === Chars.MathematicalSpace ||
    ch === Chars.IdeographicSpace ||
    ch === Chars.ThinSpace ||
    ch === Chars.ByteOrderMark
  );
}

/**
 * Converts a value to a hex value
 *
 * @param code CodePoint
 */
export function toHex(code: number): number {
  return code < Chars.UpperA ? code - Chars.Zero : (code - Chars.UpperA + 10) & 0xf;
}

/**
 * Converts a token to a string representation
 *
 * @param t Token
 */
export function convertTokenType(t: Token): string {
  switch (t) {
    case Token.NumericLiteral:
      return 'NumericLiteral';
    case Token.StringLiteral:
      return 'StringLiteral';
    case Token.FalseKeyword:
    case Token.TrueKeyword:
      return 'BooleanLiteral';
    case Token.NullKeyword:
      return 'NullLiteral';
    case Token.RegularExpression:
      return 'RegularExpression';
    case Token.TemplateContinuation:
    case Token.TemplateSpan:
    case Token.Template:
      return 'TemplateLiteral';
    default:
      if ((t & Token.IsIdentifier) === Token.IsIdentifier) return 'Identifier';
      if ((t & Token.Keyword) === Token.Keyword) return 'Keyword';

      return 'Punctuator';
  }
}
