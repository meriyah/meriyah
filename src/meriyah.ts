import { Context } from './common';
import { parseSource, Options } from './parser';
import * as ESTree from './estree';
// Current version
import { version } from '../package.json';

/**
 * Parse a script, optionally with various options.
 */
export function parseScript(source: string, options?: Options): ESTree.Program {
  return parseSource(source, options, Context.None);
}

/**
 * Parse a module, optionally with various options.
 */
export function parseModule(source: string, options?: Options): ESTree.Program {
  return parseSource(source, options, Context.Strict | Context.Module);
}

/**
 * Parse a module or a script, optionally with various options.
 */
export function parse(source: string, options?: Options): ESTree.Program {
  return parseSource(source, options, Context.None);
}

export { Options, ESTree, version };
