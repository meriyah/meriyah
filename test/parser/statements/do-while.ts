import { Context } from '../../../src/common';
import { pass, fail } from '../../test-utils';

describe('Statements - Do while', () => {
  fail('Statements - Do while (fail)', [
    ['do foo while (bar);', Context.None],
    ['do async \n f(){}; while (y)', Context.None],
    ['do async \n () => x; while(y)', Context.None],
    ['do async () \n => x; while(y)', Context.None],
    ['do let x = 1; while (false)', Context.None],
    ['do switch(x){} while(x) x', Context.None],
    ['do try {} catch {} while(x) x', Context.None],
    ['do try {} catch {} while(x) x', Context.None],
    ['do if (x) {} while(x) x', Context.None],
    ['do debugger; while(x) x', Context.None],
    ['do debugger while(x) x', Context.None]
  ]);

  pass('Statements - Do while (pass)', [
    [
      `do x
    while ({ [y]: {} ? null : false  })`,
      Context.OptionsRanges,
      {
        type: 'Program',
        start: 0,
        end: 44,
        body: [
          {
            type: 'DoWhileStatement',
            start: 0,
            end: 44,
            body: {
              type: 'ExpressionStatement',
              start: 3,
              end: 4,
              expression: {
                type: 'Identifier',
                start: 3,
                end: 4,
                name: 'x'
              }
            },
            test: {
              type: 'ObjectExpression',
              start: 16,
              end: 43,
              properties: [
                {
                  type: 'Property',
                  start: 18,
                  end: 40,
                  method: false,
                  shorthand: false,
                  computed: true,
                  key: {
                    type: 'Identifier',
                    start: 19,
                    end: 20,
                    name: 'y'
                  },
                  value: {
                    type: 'ConditionalExpression',
                    start: 23,
                    end: 40,
                    test: {
                      type: 'ObjectExpression',
                      start: 23,
                      end: 25,
                      properties: []
                    },
                    consequent: {
                      type: 'Literal',
                      start: 28,
                      end: 32,
                      value: null
                    },
                    alternate: {
                      type: 'Literal',
                      start: 35,
                      end: 40,
                      value: false
                    }
                  },
                  kind: 'init'
                }
              ]
            }
          }
        ],
        sourceType: 'script'
      }
    ],
    [
      'do async \n while (y)',
      Context.None,
      {
        body: [
          {
            body: {
              expression: {
                name: 'async',
                type: 'Identifier'
              },
              type: 'ExpressionStatement'
            },
            test: {
              name: 'y',
              type: 'Identifier'
            },
            type: 'DoWhileStatement'
          }
        ],
        sourceType: 'script',
        type: 'Program'
      }
    ],
    [
      'do async \n () \n while (y)',
      Context.None,
      {
        body: [
          {
            body: {
              expression: {
                arguments: [],
                callee: {
                  name: 'async',
                  type: 'Identifier'
                },
                type: 'CallExpression'
              },
              type: 'ExpressionStatement'
            },
            test: {
              name: 'y',
              type: 'Identifier'
            },
            type: 'DoWhileStatement'
          }
        ],
        sourceType: 'script',
        type: 'Program'
      }
    ],
    [
      'do while (x) continue \n while (x);',
      Context.None,
      {
        body: [
          {
            body: {
              body: {
                label: null,
                type: 'ContinueStatement'
              },
              test: {
                name: 'x',
                type: 'Identifier'
              },
              type: 'WhileStatement'
            },
            test: {
              name: 'x',
              type: 'Identifier'
            },
            type: 'DoWhileStatement'
          }
        ],
        sourceType: 'script',
        type: 'Program'
      }
    ],
    [
      'do foo; while (bar);',
      Context.OptionsRanges,
      {
        type: 'Program',
        start: 0,
        end: 20,
        body: [
          {
            type: 'DoWhileStatement',
            start: 0,
            end: 20,
            body: {
              type: 'ExpressionStatement',
              start: 3,
              end: 7,
              expression: {
                type: 'Identifier',
                start: 3,
                end: 6,
                name: 'foo'
              }
            },
            test: {
              type: 'Identifier',
              start: 15,
              end: 18,
              name: 'bar'
            }
          }
        ],
        sourceType: 'script'
      }
    ]
  ]);
});
