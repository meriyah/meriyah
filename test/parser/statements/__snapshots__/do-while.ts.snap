// Vitest Snapshot v1, https://vitest.dev/guide/snapshot.html

exports[`Statements - Do while > Statements - Do while (fail) > do
  a
  b
while(c); 1`] = `
"SyntaxError [3:2-3:3]: Expected 'while'
  1 | do
  2 |   a
> 3 |   b
    |   ^ Expected 'while'
  4 | while(c);"
`;

exports[`Statements - Do while > Statements - Do while (fail) > do ()=>x while(c) 1`] = `
"SyntaxError [1:9-1:14]: Unexpected token: 'while'
> 1 | do ()=>x while(c)
    |          ^^^^^ Unexpected token: 'while'"
`;

exports[`Statements - Do while > Statements - Do while (fail) > do async 
 () => x; while(y) 1`] = `
"SyntaxError [2:4-2:6]: No line break is allowed after '=>'
  1 | do async 
> 2 |  () => x; while(y)
    |     ^^ No line break is allowed after '=>'"
`;

exports[`Statements - Do while > Statements - Do while (fail) > do async 
 f(){}; while (y) 1`] = `
"SyntaxError [2:1-2:2]: Expected 'while'
  1 | do async 
> 2 |  f(){}; while (y)
    |  ^ Expected 'while'"
`;

exports[`Statements - Do while > Statements - Do while (fail) > do async () 
 => x; while(y) 1`] = `
"SyntaxError [2:1-2:3]: No line break is allowed after '=>'
  1 | do async () 
> 2 |  => x; while(y)
    |  ^^ No line break is allowed after '=>'"
`;

exports[`Statements - Do while > Statements - Do while (fail) > do debugger while(x) x 1`] = `
"SyntaxError [1:12-1:17]: Unexpected token: 'while'
> 1 | do debugger while(x) x
    |             ^^^^^ Unexpected token: 'while'"
`;

exports[`Statements - Do while > Statements - Do while (fail) > do foo while (bar); 1`] = `
"SyntaxError [1:7-1:12]: Unexpected token: 'while'
> 1 | do foo while (bar);
    |        ^^^^^ Unexpected token: 'while'"
`;

exports[`Statements - Do while > Statements - Do while (fail) > do foo while (bar); 2`] = `
"SyntaxError [1:7-1:12]: Unexpected token: 'while'
> 1 | do foo while (bar);
    |        ^^^^^ Unexpected token: 'while'"
`;

exports[`Statements - Do while > Statements - Do while (fail) > do let {} = y 1`] = `
"SyntaxError [1:7-1:8]: Unexpected token: '{'
> 1 | do let {} = y
    |        ^ Unexpected token: '{'"
`;

exports[`Statements - Do while > Statements - Do while (fail) > do let x = 1; while (false) 1`] = `
"SyntaxError [1:7-1:8]: Unexpected token: 'identifier'
> 1 | do let x = 1; while (false)
    |        ^ Unexpected token: 'identifier'"
`;

exports[`Statements - Do while > Statements - Do while (fail) > do throw function (v, h) {
  "use strict"
} while (("")) 1`] = `
"SyntaxError [3:2-3:7]: Unexpected token: 'while'
  1 | do throw function (v, h) {
  2 |   "use strict"
> 3 | } while ((""))
    |   ^^^^^ Unexpected token: 'while'"
`;

exports[`Statements - Do while > Statements - Do while (fail) > do x, y while (z) 1`] = `
"SyntaxError [1:8-1:13]: Unexpected token: 'while'
> 1 | do x, y while (z)
    |         ^^^^^ Unexpected token: 'while'"
`;

exports[`Statements - Do while > Statements - Do while (fail) > do x: function s(){}while(y) 1`] = `
"SyntaxError [1:6-1:14]: Without web compatibility enabled functions can not be declared at top level, inside a block, or as the body of an if statement
> 1 | do x: function s(){}while(y)
    |       ^^^^^^^^ Without web compatibility enabled functions can not be declared at top level, inside a block, or as the body of an if statement"
`;

exports[`Statements - Do while > Statements - Do while (pass) > do
  ()=>x
while(c) 1`] = `
{
  "body": [
    {
      "body": {
        "end": 10,
        "expression": {
          "async": false,
          "body": {
            "end": 10,
            "name": "x",
            "range": [
              9,
              10,
            ],
            "start": 9,
            "type": "Identifier",
          },
          "end": 10,
          "expression": true,
          "generator": false,
          "params": [],
          "range": [
            5,
            10,
          ],
          "start": 5,
          "type": "ArrowFunctionExpression",
        },
        "range": [
          5,
          10,
        ],
        "start": 5,
        "type": "ExpressionStatement",
      },
      "end": 19,
      "range": [
        0,
        19,
      ],
      "start": 0,
      "test": {
        "end": 18,
        "name": "c",
        "range": [
          17,
          18,
        ],
        "start": 17,
        "type": "Identifier",
      },
      "type": "DoWhileStatement",
    },
  ],
  "end": 19,
  "range": [
    0,
    19,
  ],
  "sourceType": "script",
  "start": 0,
  "type": "Program",
}
`;

exports[`Statements - Do while > Statements - Do while (pass) > do
  (function(){})
while(y) 1`] = `
{
  "body": [
    {
      "body": {
        "end": 19,
        "expression": {
          "async": false,
          "body": {
            "body": [],
            "end": 18,
            "range": [
              16,
              18,
            ],
            "start": 16,
            "type": "BlockStatement",
          },
          "end": 18,
          "generator": false,
          "id": null,
          "params": [],
          "range": [
            6,
            18,
          ],
          "start": 6,
          "type": "FunctionExpression",
        },
        "range": [
          5,
          19,
        ],
        "start": 5,
        "type": "ExpressionStatement",
      },
      "end": 28,
      "range": [
        0,
        28,
      ],
      "start": 0,
      "test": {
        "end": 27,
        "name": "y",
        "range": [
          26,
          27,
        ],
        "start": 26,
        "type": "Identifier",
      },
      "type": "DoWhileStatement",
    },
  ],
  "end": 28,
  "range": [
    0,
    28,
  ],
  "sourceType": "script",
  "start": 0,
  "type": "Program",
}
`;

exports[`Statements - Do while > Statements - Do while (pass) > do
  for((function(){});;)x
while(x); 1`] = `
{
  "body": [
    {
      "body": {
        "body": {
          "end": 27,
          "expression": {
            "end": 27,
            "name": "x",
            "range": [
              26,
              27,
            ],
            "start": 26,
            "type": "Identifier",
          },
          "range": [
            26,
            27,
          ],
          "start": 26,
          "type": "ExpressionStatement",
        },
        "end": 27,
        "init": {
          "async": false,
          "body": {
            "body": [],
            "end": 22,
            "range": [
              20,
              22,
            ],
            "start": 20,
            "type": "BlockStatement",
          },
          "end": 22,
          "generator": false,
          "id": null,
          "params": [],
          "range": [
            10,
            22,
          ],
          "start": 10,
          "type": "FunctionExpression",
        },
        "range": [
          5,
          27,
        ],
        "start": 5,
        "test": null,
        "type": "ForStatement",
        "update": null,
      },
      "end": 37,
      "range": [
        0,
        37,
      ],
      "start": 0,
      "test": {
        "end": 35,
        "name": "x",
        "range": [
          34,
          35,
        ],
        "start": 34,
        "type": "Identifier",
      },
      "type": "DoWhileStatement",
    },
  ],
  "end": 37,
  "range": [
    0,
    37,
  ],
  "sourceType": "script",
  "start": 0,
  "type": "Program",
}
`;

exports[`Statements - Do while > Statements - Do while (pass) > do
  if(x=>{});
while(y) 1`] = `
{
  "body": [
    {
      "body": {
        "alternate": null,
        "consequent": {
          "end": 15,
          "range": [
            14,
            15,
          ],
          "start": 14,
          "type": "EmptyStatement",
        },
        "end": 15,
        "range": [
          5,
          15,
        ],
        "start": 5,
        "test": {
          "async": false,
          "body": {
            "body": [],
            "end": 13,
            "range": [
              11,
              13,
            ],
            "start": 11,
            "type": "BlockStatement",
          },
          "end": 13,
          "expression": false,
          "generator": false,
          "params": [
            {
              "end": 9,
              "name": "x",
              "range": [
                8,
                9,
              ],
              "start": 8,
              "type": "Identifier",
            },
          ],
          "range": [
            8,
            13,
          ],
          "start": 8,
          "type": "ArrowFunctionExpression",
        },
        "type": "IfStatement",
      },
      "end": 24,
      "range": [
        0,
        24,
      ],
      "start": 0,
      "test": {
        "end": 23,
        "name": "y",
        "range": [
          22,
          23,
        ],
        "start": 22,
        "type": "Identifier",
      },
      "type": "DoWhileStatement",
    },
  ],
  "end": 24,
  "range": [
    0,
    24,
  ],
  "sourceType": "script",
  "start": 0,
  "type": "Program",
}
`;

exports[`Statements - Do while > Statements - Do while (pass) > do { } while (a); /^.*$/.test(b) 1`] = `
{
  "body": [
    {
      "body": {
        "body": [],
        "type": "BlockStatement",
      },
      "test": {
        "name": "a",
        "type": "Identifier",
      },
      "type": "DoWhileStatement",
    },
    {
      "expression": {
        "arguments": [
          {
            "name": "b",
            "type": "Identifier",
          },
        ],
        "callee": {
          "computed": false,
          "object": {
            "regex": {
              "flags": "",
              "pattern": "^.*$",
            },
            "type": "Literal",
            "value": /\\^\\.\\*\\$/,
          },
          "optional": false,
          "property": {
            "name": "test",
            "type": "Identifier",
          },
          "type": "MemberExpression",
        },
        "optional": false,
        "type": "CallExpression",
      },
      "type": "ExpressionStatement",
    },
  ],
  "sourceType": "script",
  "type": "Program",
}
`;

exports[`Statements - Do while > Statements - Do while (pass) > do {} while (false) false 1`] = `
{
  "body": [
    {
      "body": {
        "body": [],
        "type": "BlockStatement",
      },
      "test": {
        "type": "Literal",
        "value": false,
      },
      "type": "DoWhileStatement",
    },
    {
      "expression": {
        "type": "Literal",
        "value": false,
      },
      "type": "ExpressionStatement",
    },
  ],
  "sourceType": "script",
  "type": "Program",
}
`;

exports[`Statements - Do while > Statements - Do while (pass) > do async 
 () 
 while (y) 1`] = `
{
  "body": [
    {
      "body": {
        "expression": {
          "arguments": [],
          "callee": {
            "name": "async",
            "type": "Identifier",
          },
          "optional": false,
          "type": "CallExpression",
        },
        "type": "ExpressionStatement",
      },
      "test": {
        "name": "y",
        "type": "Identifier",
      },
      "type": "DoWhileStatement",
    },
  ],
  "sourceType": "script",
  "type": "Program",
}
`;

exports[`Statements - Do while > Statements - Do while (pass) > do async 
 while (y) 1`] = `
{
  "body": [
    {
      "body": {
        "expression": {
          "name": "async",
          "type": "Identifier",
        },
        "type": "ExpressionStatement",
      },
      "test": {
        "name": "y",
        "type": "Identifier",
      },
      "type": "DoWhileStatement",
    },
  ],
  "sourceType": "script",
  "type": "Program",
}
`;

exports[`Statements - Do while > Statements - Do while (pass) > do foo; while (bar); 1`] = `
{
  "body": [
    {
      "body": {
        "end": 7,
        "expression": {
          "end": 6,
          "name": "foo",
          "range": [
            3,
            6,
          ],
          "start": 3,
          "type": "Identifier",
        },
        "range": [
          3,
          7,
        ],
        "start": 3,
        "type": "ExpressionStatement",
      },
      "end": 20,
      "range": [
        0,
        20,
      ],
      "start": 0,
      "test": {
        "end": 18,
        "name": "bar",
        "range": [
          15,
          18,
        ],
        "start": 15,
        "type": "Identifier",
      },
      "type": "DoWhileStatement",
    },
  ],
  "end": 20,
  "range": [
    0,
    20,
  ],
  "sourceType": "script",
  "start": 0,
  "type": "Program",
}
`;

exports[`Statements - Do while > Statements - Do while (pass) > do h(function(){});while(x) 1`] = `
{
  "body": [
    {
      "body": {
        "expression": {
          "arguments": [
            {
              "async": false,
              "body": {
                "body": [],
                "type": "BlockStatement",
              },
              "generator": false,
              "id": null,
              "params": [],
              "type": "FunctionExpression",
            },
          ],
          "callee": {
            "name": "h",
            "type": "Identifier",
          },
          "optional": false,
          "type": "CallExpression",
        },
        "type": "ExpressionStatement",
      },
      "test": {
        "name": "x",
        "type": "Identifier",
      },
      "type": "DoWhileStatement",
    },
  ],
  "sourceType": "script",
  "type": "Program",
}
`;

exports[`Statements - Do while > Statements - Do while (pass) > do if(8)function s(){}
while(y) 1`] = `
{
  "body": [
    {
      "body": {
        "alternate": null,
        "consequent": {
          "async": false,
          "body": {
            "body": [],
            "type": "BlockStatement",
          },
          "generator": false,
          "id": {
            "name": "s",
            "type": "Identifier",
          },
          "params": [],
          "type": "FunctionDeclaration",
        },
        "test": {
          "type": "Literal",
          "value": 8,
        },
        "type": "IfStatement",
      },
      "test": {
        "name": "y",
        "type": "Identifier",
      },
      "type": "DoWhileStatement",
    },
  ],
  "sourceType": "script",
  "type": "Program",
}
`;

exports[`Statements - Do while > Statements - Do while (pass) > do if(8)function s(){}while(y) 1`] = `
{
  "body": [
    {
      "body": {
        "alternate": null,
        "consequent": {
          "async": false,
          "body": {
            "body": [],
            "type": "BlockStatement",
          },
          "generator": false,
          "id": {
            "name": "s",
            "type": "Identifier",
          },
          "params": [],
          "type": "FunctionDeclaration",
        },
        "test": {
          "type": "Literal",
          "value": 8,
        },
        "type": "IfStatement",
      },
      "test": {
        "name": "y",
        "type": "Identifier",
      },
      "type": "DoWhileStatement",
    },
  ],
  "sourceType": "script",
  "type": "Program",
}
`;

exports[`Statements - Do while > Statements - Do while (pass) > do if(x=>{});else n
while(y) 1`] = `
{
  "body": [
    {
      "body": {
        "alternate": {
          "end": 19,
          "expression": {
            "end": 19,
            "name": "n",
            "range": [
              18,
              19,
            ],
            "start": 18,
            "type": "Identifier",
          },
          "range": [
            18,
            19,
          ],
          "start": 18,
          "type": "ExpressionStatement",
        },
        "consequent": {
          "end": 13,
          "range": [
            12,
            13,
          ],
          "start": 12,
          "type": "EmptyStatement",
        },
        "end": 19,
        "range": [
          3,
          19,
        ],
        "start": 3,
        "test": {
          "async": false,
          "body": {
            "body": [],
            "end": 11,
            "range": [
              9,
              11,
            ],
            "start": 9,
            "type": "BlockStatement",
          },
          "end": 11,
          "expression": false,
          "generator": false,
          "params": [
            {
              "end": 7,
              "name": "x",
              "range": [
                6,
                7,
              ],
              "start": 6,
              "type": "Identifier",
            },
          ],
          "range": [
            6,
            11,
          ],
          "start": 6,
          "type": "ArrowFunctionExpression",
        },
        "type": "IfStatement",
      },
      "end": 28,
      "range": [
        0,
        28,
      ],
      "start": 0,
      "test": {
        "end": 27,
        "name": "y",
        "range": [
          26,
          27,
        ],
        "start": 26,
        "type": "Identifier",
      },
      "type": "DoWhileStatement",
    },
  ],
  "end": 28,
  "range": [
    0,
    28,
  ],
  "sourceType": "script",
  "start": 0,
  "type": "Program",
}
`;

exports[`Statements - Do while > Statements - Do while (pass) > do while (x) continue 
 while (x); 1`] = `
{
  "body": [
    {
      "body": {
        "body": {
          "label": null,
          "type": "ContinueStatement",
        },
        "test": {
          "name": "x",
          "type": "Identifier",
        },
        "type": "WhileStatement",
      },
      "test": {
        "name": "x",
        "type": "Identifier",
      },
      "type": "DoWhileStatement",
    },
  ],
  "sourceType": "script",
  "type": "Program",
}
`;

exports[`Statements - Do while > Statements - Do while (pass) > do x
while ({ [y]: {} ? null : false  }) 1`] = `
{
  "body": [
    {
      "body": {
        "end": 4,
        "expression": {
          "end": 4,
          "name": "x",
          "range": [
            3,
            4,
          ],
          "start": 3,
          "type": "Identifier",
        },
        "range": [
          3,
          4,
        ],
        "start": 3,
        "type": "ExpressionStatement",
      },
      "end": 40,
      "range": [
        0,
        40,
      ],
      "start": 0,
      "test": {
        "end": 39,
        "properties": [
          {
            "computed": true,
            "end": 36,
            "key": {
              "end": 16,
              "name": "y",
              "range": [
                15,
                16,
              ],
              "start": 15,
              "type": "Identifier",
            },
            "kind": "init",
            "method": false,
            "range": [
              14,
              36,
            ],
            "shorthand": false,
            "start": 14,
            "type": "Property",
            "value": {
              "alternate": {
                "end": 36,
                "range": [
                  31,
                  36,
                ],
                "start": 31,
                "type": "Literal",
                "value": false,
              },
              "consequent": {
                "end": 28,
                "range": [
                  24,
                  28,
                ],
                "start": 24,
                "type": "Literal",
                "value": null,
              },
              "end": 36,
              "range": [
                19,
                36,
              ],
              "start": 19,
              "test": {
                "end": 21,
                "properties": [],
                "range": [
                  19,
                  21,
                ],
                "start": 19,
                "type": "ObjectExpression",
              },
              "type": "ConditionalExpression",
            },
          },
        ],
        "range": [
          12,
          39,
        ],
        "start": 12,
        "type": "ObjectExpression",
      },
      "type": "DoWhileStatement",
    },
  ],
  "end": 40,
  "range": [
    0,
    40,
  ],
  "sourceType": "script",
  "start": 0,
  "type": "Program",
}
`;

exports[`Statements - Do while > Statements - Do while (pass) > do;while(0) 0; 1`] = `
{
  "body": [
    {
      "body": {
        "type": "EmptyStatement",
      },
      "test": {
        "type": "Literal",
        "value": 0,
      },
      "type": "DoWhileStatement",
    },
    {
      "expression": {
        "type": "Literal",
        "value": 0,
      },
      "type": "ExpressionStatement",
    },
  ],
  "sourceType": "script",
  "type": "Program",
}
`;
